### DISCUSSION TITLE
How to get macOS Ventura Spotlight to index .ipynb files?

### DISCUSSION THREAD

**Discussion by deeplook** [jupyter/notebook#6862] [Q&A]
I noticed that at some point my macOS installations stopped indexing Jupyter notebooks using the Spotlight tool. On my macOS Ventura these files are definitely not being indexed. But I do have (false?) memories that this used to work. So I did some research, but could not find any such Spotlight plugins, not to say any plugins at all. Again, I remember there were such plugins available in the past. I can also not add single file extensions in macOS Ventura. And I see almost no similar questions, in fact only [this one](https://github.com/jupyter/nbformat/issues/246). I find it very annoying to not be able to search for notebooks using Spotlight, so how do other people solve this? By using other tools like [Alfred](https://www.alfredapp.com/)?

**Comment by deeplook** [comment]
I've been pointed to a [similar question](https://discourse.jupyter.org/t/mac-osx-finder-does-not-search-ipynb-files-for-content/9905) in a Discorse forum, but also without a proper solution.

**Comment by saf-dmitry** [comment]
> I find it very annoying to not be able to search for notebooks using Spotlight, so how do other people solve this? By using other tools like [Alfred](https://www.alfredapp.com/)?

Alfred uses the same index as the Spotlight, therefore we need a proper mdimporter plugin to make the search work.

You can make the Spotlight index Jupyter notebook files as plain text JSON files, but it is not a proper solution.

**Reply by deeplook** [reply]
> You can make the Spotlight index Jupyter notebook files as plain text JSON files, but it is not a proper solution.

My understanding is that Spotlight does not index JSON files, and I don't see an importer for that format that could be tweaked to work for notebooks as well.

**Comment by saf-dmitry** [comment]
> My understanding is that Spotlight does not index JSON files, and I don't see an importer for that format that could be tweaked to work for notebooks as well.

You can modify a copy of the built-in rich text mdimporter plugin to handle Jupyter notebook files as plain text. Create a copy of the `RichText.mdimporter` package located in `/System/Library/Spotlight/` and name it as e.g., `Jupyter.mdimporter`. Inside the `Contents` folder, open the `Info.plist` file. Remove existing types and add the new file type:

```xml
<key>CFBundleDocumentTypes</key>
<array>
    <dict>
        <key>CFBundleTypeRole</key>
        <string>MDImporter</string>
        <key>LSItemContentTypes</key>
        <array>
            <string>org.jupyter.ipynb</string>
        </array>
    </dict>
</array>
```

Then put the new plugin into the `~/Library/Spotlight/` directory and run the following terminal command to reimport files claimed by the plugin:

```shell
mdimport -r ~/Library/Spotlight/Jupyter.mdimporter
```

The problem with this workaround is, that it will include not only the actual text content of notebook files but also the JSON markup. However, it's still better than no indexing at all.

**Reply by deeplook** [reply]
> You can modify a copy of the built-in rich text mdimporter plugin to handle Jupyter notebook files as plain text. Create a copy of the `RichText.mdimporter` package located in `/System/Library/Spotlight/` and name it as e.g., `Jupyter.mdimporter`. Inside the `Contents` folder, open the `Info.plist` file. Remove existing types and add the new file type: [...]

Funny, but I got a similar suggestion from ChatGPT. The [notebook format description](https://nbformat.readthedocs.io/en/latest/format_description.html) seems to define ca. 30 key names, which will clutter the desired output a little, but I guess it's still better than nothing.

To actually implement the workaround I'll need to reboot in recovery mode and disable and reenable System Integrity Protection (SIP) temporarily with `csrutil disable/enable,` which can take me some time to prepare for. In any case this is clearly a cumbersome procedure. ;-)

**Comment by saf-dmitry** [comment]
> To actually implement the workaround I'll need to reboot in recovery mode and disable and reenable System Integrity Protection 

I think, you only need to disable SIP if you want to modify the built-in rich text mdimporter plugin "in place", by adding the new format definition. But if you _copy_ the existing rich text plugin to the Spotlight user's directory and modify a copy of it from there, like I described above, there's no need to mess with SIP.

> The [notebook format description](https://nbformat.readthedocs.io/en/latest/format_description.html) seems to define ca. 30 key names, which will clutter the desired output a little

Not only that, think about embedded base-64-encoded binary content like images.

Anyway, a proper mdimporter plugin for Jupyter notebook files would be awesome.

**Reply by deeplook** [reply]
Well it seems I cannot copy anything into `/System/Library/Spotlight/` without messing with the SIP.

**Comment by saf-dmitry** [comment]
> Well it seems I cannot copy anything into `/System/Library/Spotlight/` without messing with the SIP.

You don't need to copy to `/System/Library/Spotlight/`, you need to copy to the _user_ directory `~/Library/Spotlight/`, like I described above: https://github.com/jupyter/notebook/discussions/6862#discussioncomment-5851659.

**Comment by bridgeland** [comment]
Annoying. This used to work fine, and then Ventura stopped indexing those files. Apple, have you lost your way?
